cfgov/v1/migrations/0014_grouppagepermission_index_fix.py

```diff
@@ -0,0 +1,32 @@
+# Generated by Django 3.2.23 on 2023-12-20 13:04
+
+from django.db import migrations
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ("v1", "0013_remove_cfgovpage_schema_json"),
+    ]
+
+    # This is a one-off fix to convert an existing unique index on a Wagtail
+    # table into a unique constraint. See internal Design-and-Content-Team#389.
+    #
+    # https://github.com/wagtail/wagtail/blob/b6ba78de814f9bc293e55957ac70d890d950053e/wagtail/models/__init__.py#L3060
+    table_name = "wagtailcore_grouppagepermission"
+    index_name = (
+        "idx_17388_wagtailcore_grouppagepermission_group_id_16e761a17265"
+    )
+
+    operations = [
+        migrations.RunSQL(
+            ";".join(
+                [
+                    f"DROP INDEX IF EXISTS {index_name}",
+                    f"ALTER TABLE {table_name} DROP CONSTRAINT IF EXISTS {index_name}",
+                    f"ALTER TABLE {table_name} ADD CONSTRAINT {index_name} UNIQUE(group_id, page_id, permission_type)",
```

Comment:
I think this is fine. I do wonder if it's worth considering something like a [PL/pgSQL conditional](https://www.postgresql.org/docs/13/plpgsql-control-structures.html#PLPGSQL-CONDITIONALS) around 

```pgsql
DO $$ BEGIN
    IF EXISTS (SELECT FROM pg_indexes WHERE indexname='{index_name}') THEN
        DROP INDEX {index_name};
        ALTER TABLE {table_name} ADD CONSTRAINT {index_name} UNIQUE(group_id, page_id, permission_type);
    END IF;
END $$;
```

This would avoid the need to always recreate the constraint. It does have to be wrapped in the anonymous codeblock boilerplate though.

This is just an alternative, what you have works for me.

Reply:
Thanks @willbarton, this actually works much better! It seems like the constraint name isn't always the same when I run the Django tests locally and the migrations are run during creation of the test database, so we need to do this dynamically as you suggest.